File: cpp/cpp_100-999/781.cc
// C++
// leetcode 781
// https://leetcode-cn.com/problems/rabbits-in-forest/
// 先统计回答，然后根据统计的回答与相应的频率算最少的兔子数

class Solution {
public:
    int numRabbits(vector<int>& answers) {
        map<int, int> m;
        for(auto x : answers) {
            if(m.count(x) == 0)
                m[x] = 1;
            else
                ++m[x];
        }
        int res = 0;
        for(auto it = m.begin(); it != m.end(); ++it) {
            int x = it -> first + 1;
            int times = it -> second;
            if(times % x != 0)
                times = times - (times % x) + x;
            res += times;
        }
        return res;
    }
};

File: Go/Go_100-999/781.go
// Go
// leetcode 781
// https://leetcode-cn.com/problems/rabbits-in-forest/

func numRabbits(answers []int) int {
	m := make(map[int]int)
	for _, v := range answers {
		m[v+1]++
	}

	res := 0
	for v, num := range m {
		res += ((num - 1 )/ v + 1) * v
	}

	return res
}